pulsar-zookeeper-utils/src/main/java/org/apache/pulsar/zookeeper/ZkIsolatedBookieEnsemblePlacementPolicy.java

```diff
@@ -156,6 +171,38 @@ private ZooKeeperCache getAndSetZkCache(Configuration conf) {
                 bookieToReplace, excludeBookies);
     }
 
+    private Optional<EnsemblePlacementPolicyConfig> getEnsemblePlacementPolicyConfig(
+        Map<String, byte[]> customMetadata) {
+
+        byte[] ensemblePlacementPolicyConfigData = customMetadata.get(
+            EnsemblePlacementPolicyConfig.ENSEMBLE_PLACEMENT_POLICY_CONFIG);
+        if (ensemblePlacementPolicyConfigData != null) {
+            try {
+                return Optional.ofNullable(EnsemblePlacementPolicyConfig.decode(ensemblePlacementPolicyConfigData));
+            } catch (JsonUtil.ParseJsonException e) {
+                LOG.error("Failed to parse the ensemble placement policy config from the custom metadata", e);
+                return Optional.empty();
+            }
+        }
+        return Optional.empty();
+    }
+
+    private void fillIsolationGroupWithEnsemblePlacementPolicyConfig(
+        EnsemblePlacementPolicyConfig ensemblePlacementPolicyConfig) {
+
+        String className = ZkIsolatedBookieEnsemblePlacementPolicy.class.getName();
+        if (ensemblePlacementPolicyConfig.getPolicyClass().getName().equals(className)) {
+            Map<String, Object> properties = ensemblePlacementPolicyConfig.getProperties();
+            // Because of we may have different ensemblePolicyConfig from different ledgers, so clear
+            // the isolation groups before refilling it.
+            primaryIsolationGroups.clear();
```

Comment:
Here we are modification shared data structures.
I think this is the problem that @sijie is pointing out

Reply:
Yes. I think this is one of the questions.